### PR TITLE
[SPARK-17409] [SQL] Do Not Optimize Query in CTAS More Than Once

### PR DESCRIPTION
**Pull request by gatorsmile**
### What changes were proposed in this pull request?

As explained in https://github.com/apache/spark/pull/14797:

> Some analyzer rules have assumptions on logical plans, optimizer may break these assumption, we should not pass an optimized query plan into QueryExecution (will be analyzed again), otherwise we may some weird bugs.
> For example, we have a rule for decimal calculation to promote the precision before binary operations, use PromotePrecision as placeholder to indicate that this rule should not apply twice. But a Optimizer rule will remove this placeholder, that break the assumption, then the rule applied twice, cause wrong result.

We should not optimize the query in CTAS more than once. For example, 

``` Scala
spark.range(99, 101).createOrReplaceTempView("tab1")
val sqlStmt = "SELECT id, cast(id as long) * cast('1.0' as decimal(38, 18)) as num FROM tab1"
sql(s"CREATE TABLE tab2 USING PARQUET AS $sqlStmt")
checkAnswer(spark.table("tab2"), sql(sqlStmt))
```

Before this PR, the results do not match

```
== Results ==
!== Correct Answer - 2 ==       == Spark Answer - 2 ==
![100,100.000000000000000000]   [100,null]
 [99,99.000000000000000000]     [99,99.000000000000000000]
```

After this PR, the results match.

```
+---+----------------------+
|id |num                   |
+---+----------------------+
|99 |99.000000000000000000 |
|100|100.000000000000000000|
+---+----------------------+
```

In this PR, we do not treat the `query` in CTAS as a child. Thus, the `query` will not be optimized when optimizing CTAS statement. However, we still need to analyze it for normalizing and verifying the CTAS in the Analyzer. Thus, we do it in the analyzer rule `PreprocessDDL`, because so far only this rule needs the analyzed plan of the `query`.
### How was this patch tested?

Added a test

### CODE REVIEW

**Comment by gatorsmile** [comment]
cc @cloud-fan @yhuai @davies

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/ddl.scala [hunk 1]
```diff
@@ -37,7 +38,9 @@ case class CreateTable(tableDesc: CatalogTable, mode: SaveMode, query: Option[Lo
 
   override def output: Seq[Attribute] = Seq.empty[Attribute]
 
-  override def children: Seq[LogicalPlan] = query.toSeq
+  override def children: Seq[LogicalPlan] = Seq.empty[LogicalPlan]
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/ddl.scala]: extend LeafNode?

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/ddl.scala]: Yeah. : )

**Comment by hvanhovell** [comment]
@gatorsmile so should we check all commands? It might also be an idea to have `Command` extend `LeafNode` (and make children `final`). I think @davies did something similar for https://github.com/apache/spark/pull/14797.

**Comment by gatorsmile** [comment]
@hvanhovell Sure, will do it. Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala [hunk 2]
```diff
@@ -68,7 +68,7 @@ class ResolveDataSource(sparkSession: SparkSession) extends Rule[LogicalPlan] {
 /**
  * Preprocess some DDL plans, e.g. [[CreateTable]], to do some normalization and checking.
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala]: we should update the comments to say that this rule will also analyze the query.(we may also wanna update the rule name)

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala]: Sure, let me do it now. Thanks!

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by yhuai** [comment]
@gatorsmile We should also backport this to branch 2.0, right?

**Comment by yhuai** [comment]
@gatorsmile Also, does it affect `CTAS` for creating a hive serde table?

**Comment by gatorsmile** [comment]
Yeah. We should backport it to 2.0

Yeah. It affects both data source tables and hive serde tables. To fix it in Spark 2.0, we need to rewrite the fix since Spark 2.0 does not have a unified logical plan, afaik. Let me submit a PR to backport it.

**Comment by yhuai** [comment]
Thanks! btw, does this patch cover hive tables?

**Comment by yhuai** [comment]
Also, another good test for this is 
```
val df = sql("select 0 as id")
df.registerTempTable("foo")
val df2 = sql("""select * from foo group by id""")
df2.write.mode("overwrite").saveAsTable("bar")
```
Without this fix, you will have an exception like `org.apache.spark.sql.AnalysisException: GROUP BY position 0 is not in select list (valid range is [1, 1]); line 1 pos 7`.

**Comment by yhuai** [comment]
Also, can we add a test for hive tables?

**Comment by gatorsmile** [comment]
Yeah, based on my understanding, it should cover the hive serde table. I will submit a PR to make sure it and also include the test case you provided above. Thank you!

**Pull request closed without merging**